### ISSUE TITLE
Ability to control application start order via server.xml

### ISSUE THREAD

**Issue by aguibert** [OpenLiberty/open-liberty#7331]
For a Liberty server that has multiple applications on it, we often get asked if there is a way to control the order in which applications start up. A sample use case came up recently in [this StackOverflow question](https://stackoverflow.com/questions/55834206/how-to-force-websphere-liberty-to-start-wars-in-certain-order).

Apparently @NottyCode and @brenthdaniel have discussed this recently. One possible solution they discussed would involve a `startAfter` attribute on the `<application>` element. For example:

```xml
<server>

  <application id="a" location="a.war"/>

  <application id="b" location="b.war" startAfter="a"/>
  <application id="c" location="c.war" startAfter="a"/>

  <application id="d" location="d.war" startAfter="b,c"/>

</server>
```

**Comment by tkburroughs** [comment]
EJB spec uses `@DependsOn` for this type of behavior ... slightly different way of describing the relationship.

**Comment by brenthdaniel** [comment]
Traditionally we have pushed back on this request. Certainly it will slow down server startup, and with more modern applications it would make more sense to have one app per server and orchestrate with something along the lines of docker compose. 

The implementation of something like "dependsOn" would be reasonably straightforward if we are only preventing one application from starting before another. If we expect that stopping one application will cause all dependent applications to stop as well then things get a lot trickier. Also, it probably makes sense to limit this to defined applications rather than allowing an application to depend on something from dropins.

**Comment by NottyCode** [comment]
Decision is to add `startsAfterRef` attribute that references other application's by id. We then don't start the application until the referenced application is started. This is different from other app servers which just use start order to order when the app start process is initiated. 

The `resourceAdapter` extension of applications should hide/ignore the startAfter attribute so you can't accidentally order an RA after an application in violation of the spec.

It must not be possible to define this configuration because it sucks:
```xml 
<application location="b.war">
    <startsAfter location="a.war"/>
</application>
```

This does not affect stop or restart. So stopping an application referenced by startsAfter doesn't cause other applications to stop. There is no relationship between startup order and shutdown order when the server stops.

Apps that don't specify a `startsAfterRef` start in any order in parallel. Dropins deployed apps can't be ordered. In the example blow apps `b` and `c` can start in parallel.

```xml
<server>

  <application id="a" location="a.war"/>

  <application id="b" location="b.war" startAfterRef="a"/>
  <application id="c" location="c.war" startAfterRef="a"/>

  <application id="d" location="d.war" startAfterRef="b,c"/>

</server>
```

It was discussed that we could implement an ordinal, other applications servers do this, but it would be more complicated to implement and a less direct mapping to user intent.

SIP applications may have other considerations (not an open liberty concern):

For Session Initiation Protocol (SIP) applications, the <load-on-startup> tag in the sip.xml file affects the order in which servlets within applications are started. The value that you set for Startup order on this Startup behavior console page determines the importance or weight of an application within a composition of SIP applications. For example, for the most important SIP application within a SIP application composition, specify 1 for Startup order. For the next most important SIP application within the composition, specify 2 for Startup order, and so on. For more information, see the JSR 116 specification.

**Comment by atosak** [comment]
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [ ] POC Design / WAD Review Scheduled (David Chang) or N/A.
- [ ] POC Design / WAD Reviewed (Feature Owner) or N/A.
- [ ] Complete any follow-ons from the POC Review.
- [ ] Design / WAD Approval (Alasdair Nottingham) or N/A.
- [ ] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A.
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of the epic entitled "FAT Approval Test Summary".  Add and fill in the template as described here: https://github.ibm.com/was-liberty/WS-CD-Open/wiki/Feature-Review-(Feature-Test-Summary-Process)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [ ] Implementation complete. (Epic owner / Feature owner) 
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approved or No Design Approved label on the GitHub Epic.
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Sam Wong - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
- [ ] **Demo** - (Liberty only - Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by brenthdaniel** [comment]
WAD: https://ibm.box.com/s/s6n9yxrypufjo2oehjzbyw8o6gfel12t

**Comment by gscottj** [comment]
This feature does not introduce any new or modify any existing user interfaces in ways which would require accessibility testing.

**Comment by mchenggit** [comment]
Issued captured from design review: 
- p. 11: Clarify yellow vs green mean in the diagram. 
-p 17: Consider changing startsAfter to startAfter
- Clarify behavior is enforced if all apps are restarted
- For communications: update forums such as stackoverflow, and update Request for Enhancements.
- Consider leveraging CICS use cases for early testing.

**Comment by brenthdaniel** [comment]
Design updated in box with review comments addressed:
https://ibm.box.com/s/s6n9yxrypufjo2oehjzbyw8o6gfel12t

**Comment by skasund** [comment]
No STE is required for this feature.  I've approved this feature.  Thanks!

**Comment by chirp1** [comment]
The "APPROVALS with TARGET COMPLETION DATE 2 to 1 week before Onboarding" section indicates that ID is complete or N/A. The Epic contains no "ID required" tag. Approving.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.
